### PR TITLE
Performance Improvement : Preprepare message validation

### PR DESCRIPTION
**Pull request by arc-vmware**
The validation of preprepare message although happens in a separate
bag of threads, the main thread was waiting for the validation to
happen completely.
In this commit, we are creating a separate thread which actually starts
the validations and the main thread is free to pick next set of jobs.
This should improve the throughput and cpu utilization as we are freeing
the main thread and not waiting on anything.

### CODE REVIEW

File: bftengine/src/bftengine/ReplicaImp.cpp [hunk 1]
```diff
@@ -156,6 +162,39 @@ bool ReplicaImp::validateMessage(MessageBase *msg) {
...
+ * validatePrePrepareMsg initiates the validation and waits for completion of validation.
+ * After validation, it routes the preprepare message as an internal message.
+ *
+ * @param msg : This is the PrePrepare message received by the replica, which is about
+ * to get handled.
+ * @return : returns nothing
+ */
+void ReplicaImp::validatePrePrepareMsg(PrePrepareMsg *&ppm) {
+  RequestThreadPool::getThreadPool().async(
+      [this](auto *mesg, auto *rpi, auto *iq) {
```
Comment (yuliasherman) [inline bftengine/src/bftengine/ReplicaImp.cpp:175]: mesg->msg
Parameter/variable names are unclear: rpi, iq, iMsgCvd

File: bftengine/src/bftengine/messages/AssortedInternalMsg.hpp [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+// these subcomponents is subject to the terms and conditions of the subcomponent's license, as noted in the LICENSE
+// file.
+
+#pragma once
+
+#include "ReplicaAsksToLeaveViewMsg.hpp"
+#include "PrePrepareMsg.hpp"
+
+namespace bftEngine::impl {
+struct ViewChangeIndicatorInternalMsg {
```
Comment (yuliasherman) [inline bftengine/src/bftengine/messages/AssortedInternalMsg.hpp:18]: There is no reason to keep those two messages in the same file.

File: bftengine/src/bftengine/AssortedInternalMsgCloner.hpp [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+
+#pragma once
+
+#include "messages/InternalMessage.hpp"
+#include "messages/AssortedInternalMsg.hpp"
+
+namespace bftEngine::impl {
+class AssortedInternalMsgCloner {
+ public:
+  static InternalMessage getCloneOfViewChangeIndicatorInternalMsg(ReplicaAsksToLeaveViewMsg::Reason viewChangeReason) {
```
Comment (yuliasherman) [inline bftengine/src/bftengine/AssortedInternalMsgCloner.hpp:20]: You actually not cloning any existing message, but creating a new one. Please rename the functions.
But I would actually get rid of this class at all. You could create the messages inline when needed.
There is no logic here.

File: bftengine/src/bftengine/ReplicaImp.cpp [hunk 4]
```diff
@@ -767,11 +806,10 @@ bool ReplicaImp::validatePreProcessedResults(const PrePrepareMsg *msg) {
...
   for (auto err : errors) {
     if (err) {
-      // trigger view change
-      LOG_ERROR(VC_LOG,
-                "Received PrePrepareMsg containing PreProcessResult with invalid signature: "
-                    << *err << ". Ask to leave view " << getCurrentView());
-      askToLeaveView(ReplicaAsksToLeaveViewMsg::Reason::PrimarySentBadPreProcessResult);
+      // Indicate a view change
+      LOG_WARN(VC_LOG,
+               "Received PrePrepareMsg containing PreProcessResult with invalid signature: "
```
Comment (yuliasherman) [inline bftengine/src/bftengine/ReplicaImp.cpp:811]: If touched, please use KVLOG.

File: bftengine/src/bftengine/ReplicaImp.cpp [hunk 5]
```diff
@@ -787,8 +825,10 @@ void ReplicaImp::onMessage<PrePrepareMsg>(PrePrepareMsg *msg) {
     return;
   }
 
-  if (!validatePreProcessedResults(msg)) {
-    // validatePreProcessedResults() logs the error
+  if ((!getReplicaConfig().prePrepareFinalizeAsyncEnabled) && (!validatePreProcessedResults(msg))) {
+    // trigger view change
+    LOG_ERROR(VC_LOG, "PreProcessResult Signature failure. Ask to leave " << getCurrentView());
```
Comment (yuliasherman) [inline bftengine/src/bftengine/ReplicaImp.cpp:830]: -> Ask to leave view

File: bftengine/src/bftengine/ReplicaImp.cpp [hunk 6]
```diff
@@ -1274,6 +1314,31 @@ void ReplicaImp::onInternalMsg(InternalMessage &&msg) {
...
+      return;
+    } else {
+      return onMessage(ppcim->ppm_);
+    }
+  }
+
+  // Handle a view change indicator and trigger view change
+  if (auto *vciim = std::get_if<ViewChangeIndicatorInternalMsg>(&msg)) {
+    if (bftEngine::ControlStateManager::instance().getPruningProcessStatus()) {
+      LOG_INFO(GL, "Received Viewchange indicator internal message message while pruning, so ignoring the message");
```
Comment (yuliasherman) [inline bftengine/src/bftengine/ReplicaImp.cpp:1333]: Viewchange indicator internal message -> ViewChangeIndicatorInternalMsg

File: bftengine/src/bftengine/ReplicaImp.cpp [hunk 7]
```diff
@@ -1274,6 +1314,31 @@ void ReplicaImp::onInternalMsg(InternalMessage &&msg) {
...
+      return;
+    } else {
+      // trigger view change
+      LOG_ERROR(VC_LOG, "PreProcessResult Signature failure. Ask to leave " << getCurrentView());
```
Comment (yuliasherman) [inline bftengine/src/bftengine/ReplicaImp.cpp:1337]: -> Ask to leave view

File: bftengine/src/bftengine/ReplicaImp.cpp (hunk 1 above)
Comment (arc-vmware) [reply bftengine/src/bftengine/ReplicaImp.cpp]: Thanks @yuliasherman for your comment, I have incorporated the changes.

File: bftengine/src/bftengine/messages/AssortedInternalMsg.hpp (hunk 2 above)
Comment (arc-vmware) [reply bftengine/src/bftengine/messages/AssortedInternalMsg.hpp]: Thanks @yuliasherman for your comment, I have incorporated the changes.

File: bftengine/src/bftengine/AssortedInternalMsgCloner.hpp (hunk 3 above)
Comment (arc-vmware) [reply bftengine/src/bftengine/AssortedInternalMsgCloner.hpp]: Thanks @yuliasherman for your comment, I have incorporated the changes.

File: bftengine/src/bftengine/ReplicaImp.cpp (hunk 5 above)
Comment (arc-vmware) [reply bftengine/src/bftengine/ReplicaImp.cpp]: Thanks @yuliasherman for your comment, I have incorporated the changes.

File: bftengine/src/bftengine/ReplicaImp.cpp (hunk 7 above)
Comment (arc-vmware) [reply bftengine/src/bftengine/ReplicaImp.cpp]: Thanks @yuliasherman for your comment, I have incorporated the changes.

File: bftengine/src/bftengine/ReplicaImp.cpp (hunk 4 above)
Comment (arc-vmware) [reply bftengine/src/bftengine/ReplicaImp.cpp]: Thanks @yuliasherman for your comment, I have incorporated the changes.

File: bftengine/src/bftengine/ReplicaImp.cpp (hunk 6 above)
Comment (arc-vmware) [reply bftengine/src/bftengine/ReplicaImp.cpp]: Thanks @yuliasherman for your comment, I have incorporated the changes.

File: bftengine/src/bftengine/ReplicaImp.cpp [hunk 8]
```diff
@@ -767,11 +806,10 @@ bool ReplicaImp::validatePreProcessedResults(const PrePrepareMsg *msg) {
...
-      // trigger view change
-      LOG_ERROR(VC_LOG,
-                "Received PrePrepareMsg containing PreProcessResult with invalid signature: "
-                    << *err << ". Ask to leave view " << getCurrentView());
-      askToLeaveView(ReplicaAsksToLeaveViewMsg::Reason::PrimarySentBadPreProcessResult);
+      // Indicate a view change
+      LOG_WARN(VC_LOG,
+               "Received PrePrepareMsg containing PreProcessResult with invalid signature"
+                   << KVLOG(*err) << " Replica will ask to leave view" << KVLOG(getCurrentView()));
       return false;
```
Comment (yuliasherman) [inline bftengine/src/bftengine/ReplicaImp.cpp:813]: :-) The idea was to have something like:
"Received PrePrepareMsg containing PreProcessResult with an invalid signature. The replica will ask to leave a view" << KVLOG(*err, getCurrentView()))

**Review by yuliasherman — APPROVED** [review]
LGTM, thanks!

File: bftengine/src/bftengine/ReplicaImp.cpp (hunk 8 above)
Comment (arc-vmware) [reply bftengine/src/bftengine/ReplicaImp.cpp:813]: Thanks, let me make that change and then checkin.

**Pull request merged**